packages/Teams/Middleware/SSOTokenExchangeMiddleware.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Components.Teams
+{
+    /// <summary>
+    /// If the activity name is signin/tokenExchange, this middleware will attempt to
+    /// exchange the token, and deduplicate the incoming call, ensuring only one
+    /// exchange request is processed.
+    /// </summary>
+    /// <remarks>
+    /// If a user is signed into multiple Teams clients, the Bot could receive a
+    /// "signin/tokenExchange" from each client. Each token exchange request for a
+    /// specific user login will have an identical Activity.Value.Id.
+    /// 
+    /// Only one of these token exchange requests should be processed by the bot.
+    /// The others return <see cref="System.Net.HttpStatusCode.PreconditionFailed"/>.
+    /// For a distributed bot in production, this requires a distributed storage
+    /// ensuring only one token exchange is processed. This middleware supports
+    /// CosmosDb storage found in Microsoft.Bot.Builder.Azure, or MemoryStorage for
+    /// local development. IStorage's ETag implementation for token exchange activity
+    /// deduplication.
+    /// </remarks>
+    public class SSOTokenExchangeMiddleware : IMiddleware
```

Comment:
Should we make this internal? I'm fine with registering the middleware by hooking this up as a component, but not sure if there are unintended implications if we expose this outside of the assembly otherwise.